src/MPQEditor/MPQEditor.ts

```diff
@@ -0,0 +1,647 @@
+      null,
+      this._disposables
+    );
+
+    webviewPanel.onDidDispose(() => {
+      changeDocumentSubscription.dispose();
+      while (this._disposables.length) {
+        const x = this._disposables.pop();
+        if (x) {
+          x.dispose();
+        }
+      }
+      vscode.commands.executeCommand(
+        "setContext",
+        MPQEditorProvider.viewType,
+        false
+      );
+    });
+  }
+
+  private updateWebview(
+    document: vscode.TextDocument,
+    webview: vscode.Webview
+  ): void {
+    this._mpqDataMap[document.uri.toString()].setWithString(document.getText());
+    const content = JSON.parse(document.getText());
+    if (content !== undefined) {
+      webview.postMessage({
+        type: "displayMPQ",
+        content: content,
+      });
+    }
+  }
+
+  private hadleAddSpecificLayerFromDialog(document: vscode.TextDocument) {
+    const nodes = this._mpqDataMap[document.uri.toString()].getAllModelNodes();
+    const pickOptions = {
+      canPickMany: true,
+    };
+
+    vscode.window
+      .showQuickPick(nodes, pickOptions)
+      .then((values: string | undefined) => {
+        if (!values) {
+          return;
+        }
+
+        this._mpqDataMap[document.uri.toString()].addLayers(values);
+        this.updateDocument(document);
+      });
+  }
+
+  private handleRemoveSelectedFromLayers(
+    document: vscode.TextDocument,
+    webview: vscode.Webview
+  ) {
+    let curConf = this._mpqDataMap[document.uri.toString()];
+    let selection = curConf.getSelected();
+
+    curConf.removeModelNodesFromLayers(selection);
+    curConf.clearSelection();
+
+    this.updateDocument(document);
+    webview.postMessage({
+      type: "selectionChanged",
+      names: Array<string>(),
+    });
+  }
+
+  private handleRemoveLayerFromLayers(
+    name: string,
+    document: vscode.TextDocument,
+    webview: vscode.Webview
+  ) {
+    let curConf = this._mpqDataMap[document.uri.toString()];
+
+    curConf.removeModelNodesFromLayers([name]);
+    curConf.clearSelection();
+
+    this.updateDocument(document);
+    webview.postMessage({
+      type: "selectionChanged",
+      names: Array<string>(),
+    });
+  }
+
+  private getModelFilePath(document: vscode.TextDocument): string {
+    const dirPath = path.parse(document.uri.path).dir;
+    let fileName =
+      this._mpqDataMap[document.uri.toString()].getValue("model_path");
+    return path.join(dirPath, fileName);
+  }
+
+  private handleRequestModelNodes(
+    document: vscode.TextDocument,
+    webview: vscode.Webview
+  ): void {
+    const K_DATA: string = "data";
+    const K_EXIT: string = "exit";
+    const K_ERROR: string = "error";
+    let modelFilePath = this.getModelFilePath(document);
+
+    // TODO integrate with Toolchain
+    const tool = "/usr/share/one/bin/circle-operator";
+    if (!fs.existsSync(tool)) {
+      // check whether it is installed
+      Balloon.info(
+        "To add more layers for editing please install ONE-toolchain"
```

Comment:
```suggestion
        "Please install ONE-toolchain to edit with more layers."
```

How about this re-phrasing?